plugins/woocommerce-admin/client/customize-store/design-without-ai/services.ts

```diff
@@ -45,6 +53,72 @@ const installAndActivateTheme = async () => {
 	}
 };
 
+const installFontFamilies = async () => {
```

Comment:
@gigitux I mentioned this a couple of days ago on Slack, but cross-posting here as well: Ideally, I'd recommend attaching the installation of fonts to the moment the user agreed & opted-in to our services rather than during the loading screen in the Assembler: this way we won't impact performance. We _could_ still have this here as an emergency fallback, but ideally, it should be rarely triggered (as the installation of the fonts should be successful during initial onboarding).

Reply:
>  Ideally, I'd recommend attaching the installation of fonts to the moment the user agreed & opted-in to our services rather than during the loading screen in the Assembler: this way we won't impact performance. We could still have this here as an emergency fallback, but ideally, it should be rarely triggered (as the installation of the fonts should be successful during initial onboarding).

Currently, the UI/UX for the opted-in view isn't read, and working on this pull request necessitated the installation of fonts. 
It's worth noting that installing these fonts does consume a considerable amount of time. Ideally, I believe it would be more practical to incorporate this process into the setup phase. However, engaging in any detailed discussions on this matter may be premature until we gain clarity on how the interaction should function.

Reply:
> Ideally, I believe it would be more practical to incorporate this process into the setup phase. However, engaging in any detailed discussions on this matter may be premature until we gain clarity on how the interaction should function.

The copy or placement of the use consent don't really matter in this case: what matters is where the user consent is saved on the DB to trigger an async scheduled action to install the fonts. It's totally fine if you want to combine this work with the implementation of the UX, though. Let's create an issue so we don't lose track of this task.

> It's worth noting that installing these fonts does consume a considerable amount of time.

Yep, that's precisely the reason why I'm recommending not attaching it to the CYS flow, but to the moment consent is granted (which we are planning to add right after the WooCommerce plugin is installed).

Reply:
> The copy or placement of the user consent don't really matter in this case: what matters is where (and when) the user consent is saved on the DB to trigger an async scheduled action to install the fonts. It's totally fine if you want to combine this work with the implementation of the UX, though. Let's create an issue so we don't lose track of this task.

Currently, the fonts are installed via REST API. Currently, no PHP API is available to proceed with the installation of the fonts.

All the logic is inside the REST controller: https://github.com/WordPress/gutenberg/blob/0227f403c70d96fcb65f11394cc58d74e3cf44e6/lib/experimental/fonts/font-library/class-wp-rest-font-faces-controller.php/#L284-L357

Also, all these requests are done in parallel with the creation of the products, so overall, I don't think that it increases the time of the loading.

Reply:
Another crucial aspect to consider is that, as a user, I am only agreeing to the conditions. Therefore, it seems counterintuitive to install fonts when I haven't even initiated the Customize Your Store flow.

Reply:
> Also, all these requests are done in parallel with the creation of the products, so overall, I don't think that it increases the time of the loading.

Could we make a real performance evaluation here to confirm this assumption is correct? I would rather us base this decision (whether the font installation should be preserved within the CYS onboarding flow or not) on real numbers rather than assumptions. In the end this is the primary motivation for detaching the font install from this flow. If we confirm this is accurate, then I don't have any problems with keeping the installation as is.

> Another crucial aspect to consider is that, as a user, I am only agreeing to the conditions. Therefore, it seems counterintuitive to install fonts when I haven't even initiated the Customize Your Store flow.

When the user agrees with the terms and conditions, they are agreeing to allow us to give them the best possible experience, which includes patterns, fonts, and all the good stuff. I'm not too worried, and I don't think anyone would complain about having more fonts available on their site, but we can confirm with the leadership later on if we identify there's a significant real win in performance by detaching this installation from the onboarding flow.

Reply:
> Could we make a real performance evaluation here to confirm this assumption is correct? I would rather us base this decision (whether the font installation should be preserved within the CYS onboarding flow or not) on real numbers rather than assumptions. In the end this is the primary motivation for detaching the font install from this flow. If we confirm this is accurate, then I don't have any problems with keeping the installation as is.

I already did an evaluation, but if you prefer a more in-depth check, I'm happy to do it. This is the site without products and without fonts with the current branch:


https://github.com/woocommerce/woocommerce/assets/4463174/7d4b5ebc-0bc7-4eed-8f9a-c23ae43145ff

As you can see, while all the requests regarding fonts are done, the request to generate the products is still pending.

In this video, I'm filtering the request by time. The `/products` endpoint is the last and it is long 7 seconds:
https://github.com/woocommerce/woocommerce/assets/4463174/a4ab6578-76ee-4e78-bdf9-0fadf431a24e

In the video below, I'm using trunk, and as you can see the request to `/products` endpoint is still the last and it is still long 7 seconds:

https://github.com/woocommerce/woocommerce/assets/4463174/5295fad0-5c19-4873-acd2-2fab00bb9c84

Reply:
Thanks for your response, @gigitux : 

> I already did an evaluation, but if you prefer a more in-depth check, I'm happy to do it.

Could you please share here the approx. number in seconds that all fonts are taking to load specifically? I asking this because:

> As you can see, while all the requests regarding fonts are done, the request to generate the products is still pending.

Just because a completely different endpoint is taking longer to complete (the request to the products endpoint), I don't agree that fonts should be a-ok just because they are under 7s. The request to the products endpoint is a bottleneck that we are discussing & addressing separately: we should treat each one of these requests independently, as they are.

> True, but as I said earlier, the only way to install new fonts easily is via REST API. Most of the logic is inside the REST Controllers.

> Scheduling the installation of fonts would require us to copy-paste some logic from the REST controllers into our codebase.

I see this as an opportunity for us to open a discussion & contribute upstream. We already have a number of occasions where we did copy code from GB to our plugin (e.g., Patterns registration) to achieve our dedicated goals, so I don't see this as a show-stopper.

We already have a significantly slow loading time on WooExpress, so I'd like us to avoid adding extra time to it whenever we can.

Reply:
Thank you, @nefeline! I agree with you! We should aim to the fastest onboarding experience! 🚀

> Could you please share here the approx. number in seconds that all fonts are taking to load specifically? I am asking this because:

To be honest, I'm not sure why I'm getting different results today, but I did more accurate test today and the real answer is depend:

### TT4 Installation - Activation + Products creation + Font installation

https://d.pr/v/zPttUx

```
THEME: 6182.14404296875 ms
PRODUCTS: 9698.5849609375 ms
FONTS: 11833.39892578125 ms
```
### Products Creation + Font Installation


https://d.pr/v/V1f0GD

```
THEME: 2262.531982421875 ms
PRODUCTS: 8595.154052734375 ms
FONTS: 8964.864013671875 ms
```
### TT4 Installation - Activation + Font installation

https://d.pr/v/dhnOG6

PRODUCTS: 2318.61083984375 ms
THEME: 5477.593994140625 ms
FONTS: 7278.52587890625 ms

###  Font installation

https://d.pr/v/9EA7Bm

PRODUCTS: 2318.61083984375 ms
THEME: 5477.593994140625 ms
FONTS: 7278.52587890625 ms


I agree that we should continue with the optimization, but I'm not sure that an async scheduled action is the right way to go for multiple reasons:

- There isn't PHP API for installing fonts, but as you said, we can contribute upstream and/or copy-paste the code.
- Based on my understanding, the action scheduler operates on a 60-second interval, which means there's a window of time during which fonts may not be installed. This could potentially impact some users who rely on consistent font availability. Furthermore, leveraging the scheduler for longer tasks or features that don't demand immediate user interaction, such as sending emails, can be beneficial. In my opinion, we should delay access to the assembler process until the necessary fonts are available (with the loading screen).

Reply:
Thanks for this performance evaluation for fonts, @gigitux , very useful information: even though the results may vary between 7 to 11s, it is safe to assume that this has a significant impact on performance.

> I agree that we should continue with the optimization, but I'm not sure that an async scheduled action is the right way to go for multiple reasons:
> There isn't PHP API for installing fonts, but as you said, we can contribute upstream and/or copy-paste the code.
> Based on my understanding, the action scheduler operates on a 60-second interval, which means there's a window of time during which fonts may not be installed. 

We can schedule an action to run anytime we want, it will operate whenever we schedule it to run. Take https://github.com/woocommerce/woocommerce/blob/a38c7bcccfab789e7555e9b7c34dc7d15fdd4228/plugins/woocommerce/src/Blocks/BlockPatterns.php#L330 as an example, where the action is scheduled to run immediately.

> This could potentially impact some users who rely on consistent font availability. 

Not really: if we follow an implementation as initially proposed https://github.com/woocommerce/woocommerce/pull/44004#discussion_r1466368360, with a scheduled action installing the fonts whenever the user accepts the terms & conditions combined with an "emergency" fallback font installation during the CYS loading screen to be triggered **only if** the fonts were not yet installed by the time the user reaches the CYS flow, we should cover all possible scenarios while at the same time improving performance.

Reply:
Ok, let's implement the font via async job!

Reply:
For future reference: https://github.com/woocommerce/woocommerce/issues/44257 has been created for this implementation